test/xbgpu/test_bsend.py

```diff
@@ -0,0 +1,228 @@
+from katgpucbf.utils import TimeConverter
+from katgpucbf.xbgpu.bsend import BSend
+from katgpucbf.xbgpu.output import BOutput
+
+from . import test_parameters
+
+HEAPS_PER_FENG_PER_CHUNK: Final[int] = 5
+N_TX_ITEMS: Final[int] = 2
+TOTAL_DATA_HEAPS_PER_SUBSTREAM: Final[int] = N_TX_ITEMS * HEAPS_PER_FENG_PER_CHUNK
+
+
+@pytest.fixture
+def time_converter() -> TimeConverter:
+    # TODO: Probably best to use adc_sample_rate from meerkat.py::BANDS
+    return TimeConverter(123456789.0, 1712e6)
+
+
+@pytest.fixture
+def outputs() -> Sequence[BOutput]:
+    """Simulating `--beam` configuration."""
+    return [
+        BOutput(name="foo", dst=Endpoint("239.10.11.0", 7149)),
+        BOutput(name="bar", dst=Endpoint("239.10.12.0", 7149)),
+    ]
+
+
+@pytest.fixture
+def sensors(outputs: Sequence[BOutput]) -> SensorSet:
+    """Create sensors that the send code updates."""
+    sensors = SensorSet()
+    for output in outputs:
+        sensors.add(Sensor(int, f"{output.name}.beng-clip-cnt", "Number of output samples that are saturated."))
+    return sensors
+
+
+class TestBSend:
+    """Test :class:`katgpucbf.xbgpu.bsend.BSend`."""
+
+    @staticmethod
+    async def _send_data(
+        outputs: Sequence[BOutput],
+        send_stream: BSend,
+        heap_timestamp_step: int,
+        time_converter: TimeConverter,
+        sensors: SensorSet,
+    ) -> None:
+        """Send a fixed number of heaps.
+
+        More specifically, in addition to a descriptor heap per substream, send
+        `N_TX_ITEMS` Chunks, each of which contain `HEAPS_PER_FENG_PER_CHUNK`
+        heaps.
+        """
+        # Send the descriptors as the recv_stream object needs it to
+        # interpret the received heaps correctly.
+        for i, _ in enumerate(outputs):
+            await send_stream.stream.async_send_heap(
+                send_stream.descriptor_heap,
+                substream_index=i,
+            )
+
+        for i in range(N_TX_ITEMS):
+            # Get a free chunk - there is not always a free one available. This
+            # function blocks until one is available.
+            chunk = await send_stream.get_free_chunk()
+
+            # Populate the buffer with dummy data.
+            chunk.data.fill(0)
+
+            # Give the chunk back to the send_stream to transmit out
+            # onto the network.
+            chunk.timestamp = i * HEAPS_PER_FENG_PER_CHUNK * heap_timestamp_step
+            send_stream.send_chunk(chunk, time_converter, sensors)
+        # send_heap just queues data for sending but is non-blocking.
+        # Flush to ensure that the data all gets sent before we return.
+        await send_stream.stream.async_flush()
+
+    @staticmethod
+    async def _recv_data(
+        queues: list[spead2.InprocQueue],
+        channel_offset: int,
+        n_channels_per_substream: int,
+        n_spectra_per_heap: int,
+        heap_timestamp_step: int,
+    ) -> None:
+        """Receive data transmitted from :func:`_send_data`.
```

Comment:
```suggestion
        """Receive data transmitted from :meth:`_send_data`.
```

Reply:
50/50 chance every time I do this, and I seem to get it wrong each time 😅

Reply:
My daughter also puts her shoes on the wrong feet more often than not.
Maybe it's a common human thing? 🤔